scribus/plugins/import/pdf/slaoutput.cpp

```diff
@@ -3896,3 +3927,604 @@ bool SlaOutputDev::checkClip()
+}
+
+/*code mostly taken from importodg.cpp which also supports some line styles and more fill options etc...*/
+//FIXME: This needs to be implemented based on the framework
+void SlaOutputDev::finishItem(PageItem* item)
+{
+	item->ClipEdited = true;
+	item->FrameType = 3;
+	//this requires that PoLine is set
+	//FPoint wh = getMaxClipF(&item->PoLine);
+	//item->setWidthHeight(wh.x(), wh.y());
+	//item->Clip = flattenPath(item->PoLine, item->Segments);
+	item->OldB2 = item->width();
+	item->OldH2 = item->height();
+	item->updateClip();
+	item->OwnPage = m_doc->OnPage(item);
+	//item->setFillTransparency(1.0 - state->getFillOpacity() > state->getStrokeOpacity() ? state->getFillOpacity() : state->getStrokeOpacity());
+	//item->setLineTransparency(1.0);
+}
+
+PdfTextRecognition::PdfTextRecognition()
+{
+	m_textRegions.push_back(activeTextRegion);
+	setCharMode(AddCharMode::ADDFIRSTCHAR);
+}
+
+PdfTextRecognition::~PdfTextRecognition()
+{
+}
+
+void PdfTextRecognition::addTextRegion()
+{
+	activeTextRegion = TextRegion();
+	m_textRegions.push_back(activeTextRegion);
+	setCharMode(PdfTextRecognition::AddCharMode::ADDFIRSTCHAR);
+}
+
+void PdfTextRecognition::addChar(GfxState* state, double x, double y, double dx, double dy, double originX, double originY, CharCode code, int nBytes, Unicode const* u, int uLen)
+{
+	(this->*(m_addChar))(state, x, y, dx, dy, originX, originY, code, nBytes, u, uLen);
+}
+
+bool PdfTextRecognition::isNewLineOrRegion(QPointF newPosition)
+{
+	return (activeTextRegion.collinear(activeTextRegion.lastXY.y(), activeTextRegion.textRegionLines.back().baseOrigin.y()) &&
+		!activeTextRegion.collinear(newPosition.y(), activeTextRegion.lastXY.y()))
+		|| (activeTextRegion.collinear(newPosition.y(), activeTextRegion.lastXY.y())
+			&& !activeTextRegion.isCloseToX(newPosition.x(), activeTextRegion.lastXY.x()));
+}
+
+PdfGlyph PdfTextRecognition::AddFirstChar(GfxState* state, double x, double y, double dx, double dy, double originX, double originY, CharCode code, int nBytes, Unicode const* u, int uLen)
+{
+	//qDebug() << "AddFirstChar() '" << u << " : " << uLen;	
+	PdfGlyph newGlyph = PdfTextRecognition::AddCharCommon(state, x, y, dx, dy, u, uLen);
+	activeTextRegion.glyphs.push_back(newGlyph);
+	setCharMode(AddCharMode::ADDBASICCHAR);
+
+	//only need to be called for the very first point
+	auto success = activeTextRegion.addGlyphAtPoint(QPointF(x, y), newGlyph);
+	if(success == TextRegion::FrameworkLineTests::FAIL)
+		qDebug("FIXME: Rogue glyph detected, this should never happen because the coursor should move before glyphs in new regions are added.");
+	return newGlyph;
+}
+PdfGlyph PdfTextRecognition::AddCharCommon(GfxState* state, double x, double y, double dx, double dy, Unicode const* u, int uLen)
+{
+	//qDebug() << "AddBasicChar() '" << u << " : " << uLen;
+	PdfGlyph newGlyph;
+	newGlyph.dx = dx;
+	newGlyph.dy = dy;
+
+	// Convert the character to UTF-16 since that's our SVG document's encoding
+	for (int i = 0; i < uLen; i++)

```

Comment:
this still feels strange.

it boils down to:

```cpp
newGlyph.code = (char16_t)u[uLen - 1];
```

Reply:
done. I don't know what is supposed to happen if uLen > 1